### PR TITLE
Algorithm for rejecting modification

### PR DESCRIPTION
**Pull request by aboba**
Fix for Issues https://github.com/w3c/webrtc-pc/issues/1254, https://github.com/w3c/webrtc-pc/issues/1309, https://github.com/w3c/webrtc-pc/issues/1330 and https://github.com/w3c/webrtc-pc/issues/1344

### CODE REVIEW

**Comment by aboba** [comment]
@fippo Can you review?

**Comment by fippo** [comment]
LGTM. You made the second commit when I was about to nag ;-)

File: webrtc.html [hunk 1]
```diff
@@ -1015,29 +1015,48 @@
...
+                    current <a>signaling state</a> of <var>connection</var>
+                    as described in <span data-jsep=
+                    "processing-a-local-desc processing-a-remote-desc">[[!JSEP]]</span>,
                     then reject <var>p</var> with a newly
                     <a data-link-for="exception" data-lt="create">created</a>
                     <code>InvalidStateError</code> and abort these steps.</p>
                   </li>
                   <li>
+                    Let <var>lastOffer</var> be the result returned by the
+                    last call to <code>createOffer</code>.
```
Comment (taylor-b) [inline webrtc.html]: nit: It's obvious enough to me what this means, but interpreted literally, it's unclear if `lastOffer` is an `RTCSessionDescription`, a `DOMString` or a promise.

File: webrtc.html [hunk 2]
```diff
@@ -1015,29 +1015,48 @@
...
+                  </li>
+                  <li>
+                    Let <var>lastAnswer</var> be the result returned by the
+                    last call to <code>createAnswer</code>.
+                  </li>
+                  <li>
+                    If <code><var>description</var>.type</code> is
+                    <code>offer</code> and <code><var>description</var>.sdp</code>
+                    does not match <var>lastOffer</var>,
```
Comment (taylor-b) [inline webrtc.html]: "does not match" could be loosely interpreted. If our requirement is that they're byte-for-byte identical, which is what JSEP ended up saying, we should make that more clear.

File: webrtc.html (hunk 1 above)
Comment (aboba) [inline webrtc.html]: Clarified that this means RTCSessionDescriptionInit.sdp.

File: webrtc.html (hunk 2 above)
Comment (aboba) [inline webrtc.html]: Changed to "match byte-for-byte".

File: webrtc.html [hunk 3]
```diff
@@ -1015,29 +1015,51 @@
...
+                  </li>
+                  <li>
+                    Let <var>lastAnswer</var> be the value of
+                    <code><a>RTCSessionDescriptionInit</a>.sdp</code> returned
+                    by the last call to <code>createAnswer</code>.
+                  </li>
+                  <li>
+                    If <code><var>description</var>.type</code> is
+                    <code>offer</code> and <code><var>description</var>.sdp</code>
+                    does not match <var>lastOffer</var> byte-for-byte,
```
Comment (taylor-b) [inline webrtc.html]: Is "does not match byte-for-byte" too informal? Could just say "is not equal to".

Comment (aboba) [inline webrtc.html]: Done.

File: webrtc.html [hunk 4]
```diff
@@ -1015,29 +1015,51 @@
...
-                    current <a>signaling state</a> of <var>connection</var>,
+                    current <a>signaling state</a> of <var>connection</var>
+                    as described in <span data-jsep=
+                    "processing-a-local-desc processing-a-remote-desc">[[!JSEP]]</span>,
                     then reject <var>p</var> with a newly
                     <a data-link-for="exception" data-lt="create">created</a>
                     <code>InvalidStateError</code> and abort these steps.</p>
                   </li>
                   <li>
+                    Let <var>lastOffer</var> be the value of
```
Comment (taylor-b) [inline webrtc.html]: Could be written such that `lastOffer` is an internal slot.

File: webrtc.html [hunk 5]
```diff
@@ -1015,29 +1015,51 @@
...
+                  </li>
+                  <li>
+                    If <code><var>description</var>.type</code> is
+                    <code>answer</code> or <code>pranswer</code> and
+                    <code><var>description</var>.sdp</code> is not equal
+                    to <var>lastAnswer</var>, then reject <var>p</var>
+                    with a newly <a data-link-for="exception"
+                    data-lt="create">created</a>
+                    <code>InvalidModificationError</code> and abort these steps.
+                  </li>
```
Comment (taylor-b) [inline webrtc.html]: This set of steps should only apply when setting a local description; `InvalidModificationError` shouldn't be thrown when setting a remote offer that differs from `lastOffer`.

File: webrtc.html (hunk 4 above)
Comment (aboba) [inline webrtc.html]: Yes, that would be better.

File: webrtc.html (hunk 5 above)
Comment (aboba) [inline webrtc.html]: Right.

**Comment by stefhak** [comment]
LGTM.

**Pull request merged**